clang/test/SemaCUDA/function-redclare.cu

```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fsyntax-only \
+// RUN:   -isystem %S/Inputs -verify %s
+// RUN: %clang_cc1 -triple nvptx64-nvidia-cuda -fsyntax-only \
+// RUN:   -isystem %S/Inputs -fcuda-is-device -verify %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fsyntax-only \
+// RUN:   -isystem %S/Inputs -verify=redecl -Woffload-incompatible-redeclare %s
+// RUN: %clang_cc1 -triple nvptx64-nvidia-cuda -fsyntax-only \
+// RUN:   -isystem %S/Inputs -fcuda-is-device -Woffload-incompatible-redeclare -verify=redecl %s
+
+// expected-no-diagnostics
+#include "cuda.h"
+
+__device__ void f(); // redecl-note {{previous declaration is here}}
+
+void f() {} // redecl-warning {{incompatible host/device attribute with redeclaration: new declaration is __host__ function, old declaration is __device__ function. It will cause warning with nvcc}}
```

Comment:
I'm not convinced that it's something we need to fix. As far as CUDA is concerned those are two different function overloads and it's business as usual with no warning required.

nvcc does not have a concept of attribute-based overloads, so for them a function with different attributes is a sign of a potential issue, but it's not the case for clang, IMO.

Reply:
I believe turning the warning on will never work in practice, because there will likely almost always be a ton of math functions with host and device overloads.
As I said, it is *very* common to have those and a lot of such overloads get pulled in from the headers pre-included by clang.

Reply:
IMO that looks like a nice "this is incompatible with X" warning like we have for GCC, different C/C++ versions etc., but I'm not an expert here, so maybe this isn't actually an incompatibility? FWIW neither the diagnostic nor you comment make this really clear to me. (Also I'd drop the `It will cause warning with nvcc` and make the flag something like `-Wnvcc-compat`)

Reply:
> "this is incompatible with X" warning like we have for GCC

CUDA compilation includes a lot of CUDA SDK and host headers. `<cmath>` is one of them. Compiler will *always* see a lot of implicitly `__host__` math functions declared/defined by the host headers and a lot of the same functions with `__device__` attribute declared/defined for the GPU. Whether or not user code has the same warning becomes rather moot, when *all* compilations will produce tons of these warnings. 

That said, we *may* be saved by the fact that (IIRC) compiler hides the warnings that come from the system headers.

It would be good to test it on a real CUDA compilation first. If we're guaranteed to have dozens of those warnings for every compilation, it would render the flag practically useless.

Reply:
I checked with real HIP apps and the warnings will only show up for user's code or header files. There are no warnings for host/device redeclarations in HIP or clang headers since they are included as system headers.